### ISSUE TITLE
Support for MySQL optimizer hints

### ISSUE THREAD

**Issue by martinmacko47** [knex/knex#4199]
# Environment

Knex version: n/a
Database + version: MySQL
OS: n/a

# Feature discussion / request

1. Explain what is your use case

    Sometimes we need to run a query with an [optimizer hint]. Such as:

    ```sql
    select /*+ semijoin(firstmatch, loosescan) */ * from ...
    select /*+ max_execution_time(1000) */ * from ...
    ```
    
    Please note that optimizer hits are something else than [index hints] which have different syntax and use.

2. Explain what kind of feature would support this

    AFAIK, currently there is no way how to add such optimizer hints in Knex, besides falling back to raw SQL.
    It would be usefull if there was a way how to tell Knex to add such hints to the generated query.

    The hints may be used at the beginning of each query:

    ```sql
    select /*+ ... */ ...
    insert /*+ ... */ ...
    replace /*+ ... */ ...
    update /*+ ... */ ...
    delete /*+ ... */ ...
    ```
    
    And at the beginning of subqueries:
    
    ```sql
    (select /*+ ... */ ... )
    (select ... ) union (select /*+ ... */ ... )
    (select /*+ ... */ ... ) union (select /*+ ... */ ... )
    update ... where x in (select /*+ ... */ ...)
    insert ... select /*+ ... */ ...
    ```

3. Give some API proposal, how the feature should work

    We could add a new query builder method `.addHint('semijoin(firstmatch, loosescan)')` that would add the hint after the query verb.
    
    I don't know how and if there is support of some kind of optimizer hints in other SQL dialents. So I can't tell how would this work for them.

[optimizer hint]: https://dev.mysql.com/doc/refman/5.7/en/optimizer-hints.html
[index hints]: https://dev.mysql.com/doc/refman/5.7/en/index-hints.html

**Comment by kibertoad** [comment]
@martinmacko47 Would https://github.com/knex/knex/pull/2815 allow to achieve same thing, or syntax for comments and hints is different?

**Comment by martinmacko47** [comment]
@kibertoad If I understand #2815 correctly, it adds comments **before** the verb. Like in:

```
/* comment */ select ...
```

However MySQL optimizer hits must be **directly after** the verb otherwise they don't work:

```
select /*+ semijoin(firstmatch, loosescan) */ ...
```

For instance, for select method it would need to be rendered [here between `select` and `distinctClause`].

There could be some other argument for `.comment(...)` method that would specify the comment placement. But that could become complicated as there are many places a comment may be put in a query, not just before/after the verb. IDK if there are usecases to put comments on other places.

Also a tiny detail, there is [a space between `/*` and the comment content] in #2815, which is forbidden for MySQL optimizer hits. The optimizer hist must begin with `/*+` without any spaces in between.

[here between `select` and `distinctClause`]: https://github.com/knex/knex/blob/master/lib/query/compiler.js#L213
[a space between `/*` and the comment content]: https://github.com/knex/knex/pull/2815/files#diff-52cb11bf6c7e5e1c27eff98d90816b5113c1ce7ceb067954fda4af7d0282cb26R220

**Comment by kibertoad** [comment]
@martinmacko47 Gotcha.
I think that optimizer hint functionality would definitely be useful. Would you be open to submitting a PR for that? Probably you can use #2815 as a base for how to implement it, I imagine it to be quite similar in implementation.

**Comment by martinmacko47** [comment]
@kibertoad Ok. I'll try to make a PR. It will take a week or so.

**Comment by elhigu** [comment]
I think that comment feature could be extended with options that allows to tell that the comment should be actually mysql hint.

**Comment by martinmacko47** [comment]
@elhigu I agree a universal optimizer hints API would be better. Do we know how optimizer hints are implemented in other SQL dialects? I just know that MySQL supports optimizer hints as described above, and that Postress does not support optimizer hints.

Btw, not to be confused with index hints which have different syntax and use. There can't be a single API for both optimizer and index hints.

**Comment by elhigu** [comment]
AFAIK hints are usually special comments which then proxy / server interprets. I didn't mean that specific hints should be automatically converted to different DB:s, but merely to offer same knex api method for passing passing hints to different databases...

for example:

```
.hint('hint text', { only: 'mysql' }) // optional second parameter to for example limit to which DB hint is sent
```

or something like that.

**Comment by tslater** [comment]
Just put in $20. Love to see this!

**Comment by martinmacko47** [comment]
@elhigu 
> ```
> .hint('hint text', { only: 'mysql' }) // optional second parameter to for example limit to which DB hint is sent
> ```
> 
> or something like that.

For now I didn't implement this optional second parameter as there is no other method with such parameter yet. And I don't want to invent entire new convention in the API. We may add it later.

**Comment by kibertoad** [comment]
Released in 0.21.17

**Comment by tobin224** [comment]
> I think that comment feature could be extended with options that allows to tell that the comment should be actually mysql hint. Or maybe even better would be to provide some common hint API instead of reusing comment feature...





https://stackdev.space/questions/maximizing-the-efficiency-of-mysql-update-operations
